doc/source/mutest.rst

```diff
@@ -119,5 +119,88 @@ used to check that the state of the interface actually changes.
+   print(ret)  # Prints diff:  `{'dictionary_item_removed': ["root['bar']"]}`
+
+2. All objects within any array present in ``match`` must also exist and contain
+   equivalent data (as per rule #1) within a similarly located array present in
+   the json output of ``cmd``. Array order is desregarded. For example, if:
+
+.. code-block:: python
+
+   json1 = '[{"foo":"foo"}]'
+   json2 = '[{"foo":"foo"}, {"bar":"bar"}]'
+   json3 = '[{"bar":"bar"}, {"foo":"foo"}]'
+
+   # Then, the following results are observed:
+
+   _, ret = match_step_json("r1", f"echo '{json2}'", json1)  # Successfully matches!
+   print(ret)  # Prints match: `[{'foo': 'foo'}, {'bar': 'bar'}]`
+   _, ret = match_step_json("r1", f"echo '{json1}'", json2)  # Fails to match...
+   print(ret)  # Prints diff:  `{'iterable_item_removed': {'root[1]': {'bar': 'bar'}}}`
+   _, ret = match_step_json("r1", f"echo '{json2}'", json3)  # Successfully matches!
+   print(ret)  # Prints match: `[{'foo': 'foo'}, {'bar': 'bar'}]`
+
+3. All other data within an array present in either ``match`` or the json
+   output of ``cmd`` must also exist in the other json and be of equal value.
+   Array order is disregarded. For example, if:
+
+.. code-block:: python
+
+   json1 = '["foo"]'
+   json2 = '["foo", "bar"]'
+   json3 = '["bar", "foo"]'
+
+   # Then, the following results are observed:
+
+   _, ret = match_step_json("r1", f"echo '{json2}'", json1)  # Fails to match...
+   print(ret)  # Prints diff:  `{'iterable_item_added': {'root[1]': 'bar'}}`
+   _, ret = match_step_json("r1", f"echo '{json1}'", json2)  # Fails to match...
+   print(ret)  # Prints diff:  `{'iterable_item_removed': {'root[1]': 'bar'}}`
+   _, ret = match_step_json("r1", f"echo '{json2}'", json3)  # Successfully matches!
+   print(ret)  # Prints match: `['foo', 'bar']`
+
+These rules apply no matter what ``level`` of the array (``list``) or object
+(``dict``). For example, if:
+
+.. code-block:: python
+
+   json1 = '{"level1": ["level2", {"level3": ["level4"]}]}'
+   json2 = '{"level1": ["level2", {"level3": ["level4"], "l3": "l4"}]}'
+   json3 = '{"level1": ["level2", {"level3": ["level4", {"level5": "l6"}]}]}'
+   json4 = '{"level1": ["level2", {"level3": ["level4", "l4"]}]}'
+
+   # Then, the following results are observed:
+
+   _, ret = match_step_json("r1", f"echo '{json2}'", json1)  # Successfully matches (rule #1)!
+   print(ret)  # Match: `{'level1': ['level2', {'level3': ['level4'], 'l3': 'l4'}]}`
+   _, ret = match_step_json("r1", f"echo '{json3}'", json1)  # Successfully matches (rule #2)!
+   print(ret)  # Match: `{'level1': ['level2', {'level3': ['level4', {'level5': 'l6'}]}]}`
+   _, ret = match_step_json("r1", f"echo '{json4}'", json1)  # Fails to match (rule #3)...
+   print(ret)  # Diff:  `{'iterable_item_added': {"root['level1'][1]['level3'][1]": 'l4'}}`
+
+This behavior can be overrided to perform an exact json match if ``exact_match``
+is set to ``True``.
+
```

Comment:
If an exact match is desired, then ``exact_match`` should be set to ``True``.
When ``exact_match`` is ``True`` the comparison will only succeed when all data
in ``match`` is present in the ``cmd`` result *and* all data present in the
``cmd`` result is present in ``match``. In other words they exactly match.